### ISSUE TITLE
GPU utilization / Windows Build v2026.8.28

### ISSUE THREAD

**Issue by thetax2** [harry7557558/spirula-studio#31]
Attached is the Deep Research report. It can come across as quite critical in places, so I just wanted to apologize in advance. :-)
v2026.8.28

"During the training process on a massive dataset of nearly 6,000 high-resolution shots (exported in 4K and downscaled to half resolution in Spirula Studio), the  remains at a disproportionately low level of around 44 to 50 percent when the application window is maximized.

This underutilization is accompanied by a severe latency of over 500 milliseconds per training step. Simply disabling the "Live Mode" within the Graphical User Interface (GUI) yields no measurable performance increase. However, as soon as the user completely minimizes (iconifies) the application window, the GPU utilization instantly jumps to 100 percent, and the processing speed normalizes drastically."

https://gemini.google.com/share/2d526dbd092b?skid=a8018b5c-d8d3-46d5-9eb3-981fc926f537

**Comment by harry7557558** [comment]
Hi, thanks for opening the issue. I could not reproduce it on my end - during training GPU util is generally above 95% for me, on both Linux and Windows. A few questions that may help diagnosing:
- How does GPU util change if you change view from "3D view" to "Images" on top right of the training window?
- Do you have a game controller connected? If yes, how does GPU util change if you disconnect it?
- What's your display resolution? How does GPU util change if you resize window to a much smaller one, but not completely minimize it?

Let me know your answers to these questions, so I can look into further.

I did looked into the attached analysis - I believe it's not the main issue, since imgui is running on OpenGL (rather than Vulkan) backend, so there's no `VK_PRESENT_MODE_FIFO_KHR`; and training already runs on its own thread.

**Comment by thetax2** [comment]
How does GPU util change if you change view from "3D view" to "Images" on top right of the training window?
- No change to the utilization. it jumps from 54%-60%

Do you have a game controller connected? If yes, how does GPU util change if you disconnect it?
- Keyboard + mouse. But I have Nvidia GameStream running in the background, using Sunshine as the server and Moonlight as the client. The workstations are running in headless mode. It’s possible that Moonlight could be causing some issues here, since it also installs gamepad drivers. I’ll test that tomorrow.

What's your display resolution? How does GPU util change if you resize window to a much smaller one, but not completely minimize it? 
-  1080p at 144 Hz. I made an interesting discovery: at 144 Hz, the GPU load is around 51 to 60%. If I switch the refresh rate to 60 Hz in Windows, the load drops to 32 to 36%. But as soon as I move the Spirulae window around, the GPU load jumps back up to 50%+.

Resizing the window doesn’t cause any change.

**Comment by harry7557558** [comment]
Thanks for information. While I still cannot verify on my end since I cannot reproduce it, I've made a commit (referenced above) that should have something to do with GPU util. Try the latest master branch and let me know if it makes any difference for you - I've attached a binary [here](https://drive.google.com/file/d/1gLVhzL7vYVJDwF30PrPYssHF7S_2IVAE/view).

**Comment by thetax2** [comment]
Hi Harry, thanks for sharing the Master build:
https://drive.google.com/file/d/1gLVhzL7vYVJDwF30PrPYssHF7S_2IVAE/view

Unfortunately, this version runs even slower on the RTX 5090. GPU utilization is at 20%...

with the window open! If I minimize it now, GPU utilization goes up to 40%.

**Comment by harry7557558** [comment]
Hi, sorry to hear that - A few questions:
- I did a bit of research, my understanding is on some (but not all) Windows systems, task manager separates "3D" from actual compute - click top left of the "3D" and switch it to "Compute" (not available on all systems; on my system it's merged with "3D"). Have you tried checking "Compute", and is it 100% there? If yes, this means it's a problem with how task manager reports GPU util, not because Spirula Studio is under-utilizing GPU resources. You can additionally get more detailed GPU util breakdown with dedicated tools like nvidia-smi (or a wrapper like nvitop/nvtop).
    - Furthermore, if you can verify that there's no/little difference training speed (ms/step, or end-to-end training time), it confirms it's not under-utilizing GPU resources.
- Try [this build](https://drive.google.com/file/d/1gqy8hXhd9sYfK7FxJYO9YakubQRgelgT/view) (from 02077056294e0538f7ff0f4446934f43d26ae944). Does it change anything for you?
- If you disable Sunshine/Moonlight gamepad driver, does it change anything?
- If you load a dataset where images are already downscaled (instead of downscaling them on-the-fly), does it change anything?

My top suspicion is it's an issue with GPU util reporting done by task manager. Regarding 500ms/step being slow, it should note that Spirula counts steps differently from how LFS/Brush counts - Spirula traines with batch size ceil(num images / max batches per epoch) where default value for "max batches per epoch" is 800 (configurable), so for 6000 images I expect the actual ms/step to be divided by 8 - i.e. slightly over 60 ms/step in LFS/Brush standard; and in Spirula there's generally no need to increase number of training steps for datasets with large number of images. Nevertheless, 60 ms/step with 2k images on a 5090 does sound a bit slow to me. Let me know if 60 ms/step is comparable to what you get with other splatting software. If it's not the case, I can investigate further.

**Comment by harry7557558** [comment]
Hi, since I believe this has been investigated and answered in #49, I'm going to close this issue. Feel free to submit a new issue if you have any question, or let me know if you want this issue to be reopened.
